### PR TITLE
[diffusion] pipeline: lightweight warmup, denoising stage only, 1-step

### PR DESCRIPTION
**Pull request by tom-jerr**
## Motivation

lightweight warmup, denoising stage only, 1-step #13692

## Modifications
**New Parameter Control:**

- Added the --enable-warmup parameter in server_args.py.

- The default value is False. Users can explicitly enable the warmup function via this parameter.

**Implemented Warmup Logic for Denoising Stages:** 
- The _warmup method has been implemented in the following three core Denoising Stage classes. It is conditionally invoked before the forward pass based on the parameter settings.

  - DenoisingStage (denoising.py)
  
  - DmdDenoisingStage (denoising_dmd.py)

**⚠️ Attention:**
I found that for the FastWan2.1-T2V-1.3B-Diffusers video generation model, enabling warm-up without enabling torch compile results in poor performance.

I think that **warm-up reduces the average execution time of each subsequent step, but since the denoising stage has too few steps (3 steps), the cost of adding a warm-up outweighs the benefits it brings.**

**Enable torch compile:**
from #13641, just use the max-autotune-no-cudagraphs compile mode.

## Benchmarking and Profiling

All tests were performed on the H100. I fixed code to enable torch.compile.(most code from #13641). 

Now I just tested baseline and warm up.

#### 0. Denoising Stage Result

  | Baseline | Warm Up
-- | -- | --
Time per denoising step in torch profile | 727ms | 485ms
E2E latency per denoising step | 1.0839s | 0.7023s

- **Baseline**  
image

- **Warm Up**  
image

---

#### 1. High-level Summary

| Metric        | Baseline      | New          | Diff                        | Status |
|---------------|---------------|--------------|------------------------------|--------|
| E2E Latency   | 64487.07 ms   | 44302.05 ms  | -20185.03 ms (-31.3%)       | ✅     |
| Throughput    | 0.02 req/s    | 0.02 req/s   | -                            | -      |

---

#### 2. Stage Breakdown

| Stage Name              | Baseline (ms) | Warm Up (ms) | Diff (ms) | Diff (%) |
|-------------------------|----------------|--------------|-----------|----------|
| InputValidationStage    | 0.24           | 0.06         | -0.18     | -74.8%   |
| TextEncodingStage       | 1484.17        | 1537.20      | +53.03    | +3.6%    |
| ConditioningStage       | 0.04           | 0.01         | -0.03     | -70.5%   |
| TimestepPreparationStage| 59.99          | 25.27        | -34.72    | -57.9%   |
| LatentPreparationStage  | 0.58           | 0.26         | -0.32     | -54.7%   |
| DenoisingStage          | 62248.00       | 41786.81     | -20461.18 | -32.9%   |
| DecodingStage           | 669.70         | 942.67       | +272.97   | +40.8%   |

---

### FastVideo / FastWan2.1-T2V-1.3B-Diffusers Results

> **Warm-up reduces the average execution time of each subsequent step, but since the denoising stage has too few steps (3 steps), the cost of adding a warm-up outweighs the benefits it brings.**

#### 0. Denoising Stage Result

  | Baseline | Warm Up
-- | -- | --
E2E latency per denoising step | 2.2530s | 1.3381s

---

#### 1. High-level Summary

| Metric        | Baseline      | New          | Diff                        | Status |
|---------------|---------------|--------------|------------------------------|--------|
| E2E Latency   | 15062.46 ms   | 15955.15 ms  | +892.69 ms (+5.9%)          | ❌     |
| Throughput    | 0.07 req/s    | 0.06 req/s   | -                            | -      |

---

#### 2. Stage Breakdown

| Stage Name              | Baseline (ms) | Warm Up (ms) | Diff (ms) | Diff (%) |
|-------------------------|----------------|--------------|-----------|----------|
| InputValidationStage    | 0.05           | 0.06         | +0.01     | +16.1%   |
| TextEncodingStage       | 4445.95        | 4705.35      | +259.40   | +5.8%    |
| ConditioningStage       | 0.02           | 0.02         | +0.00     | +30.2%   |
| TimestepPreparationStage| 32.56          | 29.05        | -3.52     | -10.8%   |
| LatentPreparationStage  | 0.88           | 0.97         | +0.10     | +11.2%   |
| DenoisingStage          | 6123.29        | 7070.25      | +946.96   | +15.5%   |
| DecodingStage           | 4452.30        | 4141.41      | -310.89   | -7.0%    |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

**Comment by mickqian** [comment]
Could you also make sure, if it's necessary to expand the warm-up to all stages? Thanks!

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 1]
```diff
@@ -1097,6 +1168,32 @@ def prepare_extra_func_kwargs(self, func, kwargs) -> dict[str, Any]:
...
+                "encoder_hidden_states_image",
+                "mask_strategy",
+                "encoder_hidden_states_2",
+                "encoder_attention_mask",
+            }
+        for k, v in kwargs.items():
+            if k in accepted_params:
+                extra_step_kwargs[k] = v
+        return extra_step_kwargs
         import functools
```
Comment (zcnrex) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Is this expected to bypass everything below in this function?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 2]
```diff
@@ -965,6 +1032,10 @@ def forward(
         trajectory_timesteps: list[torch.Tensor] = []
         trajectory_latents: list[torch.Tensor] = []
 
+        # Warmup
+        if server_args.enable_warmup:
```
Comment (zcnrex) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Curious to learn: why does it require warm up for every request?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 1 above)
Comment (tom-jerr) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: This is a bug, I'll fix it.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 2 above)
Comment (tom-jerr) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: It is not necessary. I've already modified it.

**Comment by mickqian** [comment]
We might be considering no torch-compile for now.

>I found that for the FastWan2.1-T2V-1.3B-Diffusers video generation model, enabling warm-up without enabling torch compile results in poor performance; however, enabling --enable-torch-compile

How is that, could you do multiple tests and dump the perf comparison report?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 3]
```diff
@@ -152,11 +144,30 @@ def __init__(
...
+            return module
+        try:
+            import torch._inductor.config as _inductor_cfg
+
+            _inductor_cfg.reorder_for_compute_comm_overlap = True
+        except ImportError:
+            pass
+        mode = os.environ.get("SGLANG_TORCH_COMPILE_MODE", "max-autotune-no-cudagraphs")
+        compiled_forward = torch.compile(getattr(module, "forward"), mode=mode)
+        setattr(module, "forward", compiled_forward)
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:136]: You can add log records explaining the used compilation mode.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/causal_denoising.py [hunk 4]
```diff
@@ -180,7 +180,7 @@ def forward(
                         **pos_cond_kwargs,
                     )
                 current_start_frame += 1
-                remaining_frames = input_frames - 1
+                input_frames -= 1
```
[the remark this answers is not in the document]
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/causal_denoising.py]: It's a bug?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 3 above)
Comment (tom-jerr) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:136]: I'll do this and fix bug below.

**Review by BBuf — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 5]
```diff
@@ -904,6 +917,61 @@ def post_forward_for_ti2v_task(
 
         return latents
 
+    def _warmup(self, batch: Req, server_args: ServerArgs, prepared_vars: dict):
+        logger.info("Performing 1-step warmup")
+
+        target_dtype = prepared_vars["target_dtype"]
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: I would suggest doing an extra step here, instead of duplicating the main denoising loop

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 6]
```diff
@@ -304,6 +304,26 @@ def _maybe_enable_cache_dit(self, num_inference_steps: int) -> None:
         self._cache_dit_enabled = True
         self._cached_num_steps = num_inference_steps
 
+    def torch_compile_module(self, module):
+        """
+        Compile a module's forward with torch.compile, and enable inductor overlap tweak if available.
+        No-op if torch compile is disabled or the object has no forward.
+        """
+        if not self.server_args.enable_torch_compile or module is None:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: where does the `_maybe_enable_cache_dit` go?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 5 above)
Comment (tom-jerr) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: OK, I'll fix it.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 6 above)
Comment (tom-jerr) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Maybe there is something wrong, I'll fix it.

**Comment by mickqian** [comment]
/gemini please carefully review this PR, this PR is trying to do a complete and thorough warmup that in actual forward, there's no compile or warmup completely

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/causal_denoising.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/causal_denoising.py]: please fix it

**Comment by fsygd** [comment]
> I found PR [#15773](https://github.com/sgl-project/sglang/pull/15773) warms up by doing multiple rounds generating before actually generating.
> 
> Does that method conflict with the denoising stage lightweight warm-up? I’m a bit confused right now — which operations are we actually warming up for?@mickqian @yhyang201

I'll find you by wechat.

**Review by mickqian — APPROVED** [review]

**Comment by tom-jerr** [comment]
> @tom-jerr I've made some updates to this PR.
> 
> To sum up, now we insert an identical warmup req with `--num-inference-steps=1`, before actual req

I think I understand your point now, and I’ll rerun the tests based on these changes.

I’ve discussed this with fsygd, and I‘m wondering whether the current implementation can be orthogonal to PR https://github.com/sgl-project/sglang/pull/15773, or if we should discuss this further.

**Comment by mickqian** [comment]
duplicate with #16213
